reflect_abi_patch.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"maps"
+	"os"
+	"slices"
+	"strings"
+)
+
+func abiNamePatch(path string) (string, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return "", err
+	}
+
+	find := `return unsafe.String(n.DataChecked(1+i, "non-empty string"), l)`
+	replace := `name := unsafe.String(n.DataChecked(1+i, "non-empty string"), l)
+	return _realName(name)`
+
+	str := strings.Replace(string(data), find, replace, 1)
+
+	realname := `
+//go:linkname _realName
+func _realName(name string) string
+`
+
+	return str + realname, nil
+}
+
+var patchMainFile = ""
+
+func reflectMainPrePatch(path string) ([]byte, error) {
+	if patchMainFile != "" {
+		// already patched another file in main
+		return nil, nil
+	}
+
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	patchMainFile = path
+
+	nameMap := "\nvar _nameMap = map[string]string{}"
+
+	return append(content, []byte(realNameCode+nameMap)...), nil
+}
+
+func reflectMainPostPatch(file []byte, lpkg *listedPackage) []byte {
+	obfMapName := hashWithPackage(lpkg, "_nameMap")
+	nameMap := fmt.Sprintf("%s = map[string]string{", obfMapName)
+
+	values := ""
+	obfToReal := realForReflected()
+
+	keys := slices.Collect(maps.Keys(obfToReal))
+	slices.Sort(keys)
+
+	for _, obf := range keys {
+		values += fmt.Sprintf(`"%s": "%s",`, obf, obfToReal[obf])
+	}
+
+	return bytes.Replace(file, []byte(nameMap), []byte(nameMap+values), 1)
+}
+
+// this may look unnessecary for simpler cases but we need support complex cases like:
+// name := "*struct { AQ45rr68K string; ipq5aQSIqN string; hNfiW5O5LVq struct { gPTbGR00hu string } }"
+const realNameCode = `
+//go:linkname _realName internal/abi._realName
+func _realName(name string) string {
+	for i := 0; i < len(name); {
+		remLen := len(name[i:])
+
+		found := false
+		for obfName, real := range _nameMap {
+			keyLen := len(obfName)
+			if keyLen > remLen {
+				continue
+			}
+			if name[i:i+keyLen] == obfName {
+				name = name[:i] + real + name[i+keyLen:]
+
+				found = true
+				i += len(real)
+				break
+			}
+		}
+
+		if !found {
+			i++
+		}
+	}
+
+	return name
```

Comment:
I don't follow why this func body needs to be so complex; the godoc gives an example but it's not immediately obvious to me what that means. can you add more docs or comments here?

Reply:
The "name" from "internal/abi" can be any of the following:

"SimpleName"

"struct {fieldName string anotherField struct {nestedField string}}"

"*pkgName.FuncName"

This code is doing a more efficient/ without imports version of:
```go
for obfName, real := nameMap {
    name = strings.ReplaceAll(name, obfName, real)
}
```

Reply:
I would rather not do that as I would also like it to be possible to obfuscate just some fields of a struct (in the future).

Maybe instead we could make all the names the same length, then it would just be a map lookup at every position.

Also packages like encoding/json cache this information so it's not that bad.

Reply:
interesting idea, but the varying lengths was done on purpose to make detecting garbled binaries a bit harder, so I'd rather not undo that, even if it does help the internal/abi code. we can probably tackle the slowness here separately.